p2p/ipld/plugin/plugin.go

```diff
@@ -0,0 +1,329 @@
+package main
+
+import (
+	"bufio"
+	"bytes"
+	"crypto/sha256"
+	"errors"
+	"fmt"
+	"io"
+
+	blocks "github.com/ipfs/go-block-format"
+	"github.com/ipfs/go-cid"
+	"github.com/lazyledger/nmt"
+
+	"github.com/ipfs/go-ipfs/core/coredag"
+	"github.com/ipfs/go-ipfs/plugin"
+	format "github.com/ipfs/go-ipld-format"
+	node "github.com/ipfs/go-ipld-format"
+	mh "github.com/multiformats/go-multihash"
+)
+
+const (
+	// NMT is the codec used for this plugin.
+	// 0x77 seems to be free:
+	// https://github.com/multiformats/multicodec/blob/master/table.csv
+	NMT = 0x77
+	// DagParserFormatName can be used when putting into the IPLD Dag
+	DagParserFormatName = "extended-square-row-or-col"
+
+	// FIXME: These are the same as types.ShareSize and types.NamespaceSize.
+	// Repeated here to avoid a dependency to the wrapping repo as this makes
+	// it hard to compile and use the plugin against a local ipfs version.
+	namespaceSize = 8
+	shareSize     = 256
+)
+
+// Plugins is an exported list of plugins that will be loaded by go-ipfs.
+var Plugins = []plugin.Plugin{
+	&LazyLedgerPlugin{},
+}
+
+var _ plugin.PluginIPLD = &LazyLedgerPlugin{}
+
+type LazyLedgerPlugin struct{}
+
+func (l LazyLedgerPlugin) RegisterBlockDecoders(dec format.BlockDecoder) error {
+	dec.Register(NMT, NmtNodeParser)
+	return nil
+}
+
+func (l LazyLedgerPlugin) RegisterInputEncParsers(iec coredag.InputEncParsers) error {
+	iec.AddParser("raw", DagParserFormatName, DataSquareRowOrColumnRawInputParser)
+	return nil
+}
+
+func (l LazyLedgerPlugin) Name() string {
+	return "LazyLedger"
+}
+
+func (l LazyLedgerPlugin) Version() string {
+	return "0.0.0"
+}
+
+func (l LazyLedgerPlugin) Init(env *plugin.Environment) error {
+	return nil
+}
+
+// DataSquareRowOrColumnRawInputParser reads the raw shares and extract the IPLD nodes from the NMT tree.
+// Note, to parse without any error the input has to be of the form:
+//
+// <share_0>| ... |<share_numOfShares>
+//
+// To determine the share and the namespace size the constants
+// types.ShareSize and types.NamespaceSize are used.
```

Comment:
Instead of this, we could also encode the namespace size and the share size into what is fed into the plugin. That would be more general. For now, I didn't want to define a protobuf message, neither another encoding. We can simply assume that the namespace size, as well as the share size, are fixed / const.